Feature/port to agl

mkorniichuk opened the pull request:
This PR is **ready** for review.

### Summary
Current PR includes changes which allow to use cross-compilation in order to build SDL for AGL-based platform

### Changelog
##### Breaking Changes
* Third party structural and logical changes
* Toolchain support
* Toolchain for AGL

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

mkorniichuk committed:
Third party sources moved

Third party sources moved to corresponding directories

mkorniichuk committed:
APR library

* Custom Find APR module
* Separated build modules for AGL and Linux

mkorniichuk committed:
APR Util library

* Custom Find APR-Util
* Separated build modules for AGL and Linux
* Fixed cmake for new structure

mkorniichuk committed:
Log4CXX library

* Custom Find Log4CXX module
* Separated build modules for AGL and Linux
* Bash code for versioning moved to helpers.sh
* CMakeLists adjusted for new structure

mkorniichuk committed:
BSON library

* New Find BSON module
* Separated build modules for AGL and Linux
* CMakeLists adjusted for new structure

mkorniichuk committed:
Boost library

* Custom Find Boost module
* Separated build modules for AGL and Linux
* CMakeLists adjusted for new structure

mkorniichuk committed:
Third party cmake

* CMakeLists adjusted for new structure
* bash code moved to log4cxx helpers
* minor changes in find_paths

mkorniichuk committed:
CMakeLists adjustment for AGL build support

* added toolchain for AGL
* main CMakeLists adjustment for cross-compilation for AGL and for Linux
* minor CMakeLists issues fixed due to new third party src structure and AGL support
* fixed intergen build for cross compilation (tools)
* fixed dependency for components/utils (not just Boost but 3rd_party)
* now the message with installation directories for third party is promped while
executing cmake
* expat build options are now with 'EXPAT_' prefix and marked as advanced
* fixed helpers.sh for logger and extended docstring
* fixed logger versioning  and variable names. Removed unused variables
* fixed boost include directory
* force rebuild apr and apr-util if logger is not found or out of date
* fixed components dependencies for multithreading build

mkorniichuk committed:
Docker image for AGL and QNX build support

Dockerfiles and helpers scripts for building an image for building
SDL for AGL and QNX

LuxoftAKutsan commented on src/3rd_party/apache-log4cxx/helpers.sh:
```diff
@@ -0,0 +1,103 @@
+}
+
+
+      #   COMMAND /bin/bash -c \"cd ${CMAKE_CURRENT_SOURCE_DIR} &&
```
@mkorniichuk old code?

mkorniichuk replied:
@LuxoftAKutsan fixed within bf0ee058043a3c5988f9ca2f28ecf5c31183842f

LuxoftAKutsan commented on src/3rd_party/apache-log4cxx/helpers.sh:
```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
+
+#
```
@mkorniichuk please add to comment main functions and required variables, and some usage examples

mkorniichuk replied:
@LuxoftAKutsan fixed within bf0ee058043a3c5988f9ca2f28ecf5c31183842f

AByzhynar replied:
@mkorniichuk I would append with helper command line options `-h`, `--help` to get help

mkorniichuk replied:
@AByzhynar cmake will use this script, not a developer, so a simple comment is enough.

LuxoftAKutsan commented on src/3rd_party/apache-log4cxx/helpers.sh:
```diff
@@ -0,0 +1,103 @@
+    LOG4CXX_SOURCE_DIRECTORY=$1
+    LOG4CXX_LIBS_DIRECTORY=$2
+    COMMIT_HASH_FILE=/tmp/commit_hash
+    cd $LOG4CXX_SOURCE_DIRECTORY && git log --pretty=format:%H -1 . 1>$COMMIT_HASH_FILE 2>/dev/null
```
@mkorniichuk do we rely need to mute errors?

mkorniichuk replied:
@LuxoftAKutsan fixed within bf0ee058043a3c5988f9ca2f28ecf5c31183842f

LuxoftAKutsan commented on tools/infrastructure/docker/AGL/workspace/Dockerfile:
```diff
@@ -0,0 +1,82 @@
+FROM ubuntu:16.04
```
@mkorniichuk I would avoid committing workplace for AGL in repo

mkorniichuk replied:
@LuxoftAKutsan fixed within 472c3a9a75a1670991f453d0e6453cb371e03e3b

LuxoftAKutsan reviewed (approved):
:+1: 
Please format all your **new** changes with cmake-format

mkorniichuk committed:
Copyright fixes

copyrights are introduced to new CMake related files

mkorniichuk committed:
Gold linker fixed

USE_GOLD_LD worked for any linker. Now gold ld check improved and
USE_GOLD_LD flag is set to off by default

LuxoftAKutsan commented on CMakeLists.txt:
```diff
@@ -290,9 +282,8 @@ if(ENABLE_GCOV)
-set(CMAKE_CXX_FLAGS_DEBUG   " -DDEBUG -g3 -ggdb3")
-
+# TODO(AK): check current OS here
+add_definitions(-DOS_POSIX) # see line :360
```
@mkorniichuk line 360 : 
```
if (HMI STREQUAL "web" AND NOT DEFINED HMIADAPTER)
```
Please avoid line numbers in comments

mkorniichuk replied:
@LuxoftAKutsan fixed within a89e4042f9a29e6b3ec306046c57bd2bd2e2c4f7

LuxoftAKutsan commented on agl_linux_x86_64.cmake:
```diff
@@ -0,0 +1,35 @@
+
+include(/opt/agl-sdk/6.0.2-corei7-64/sysroots/x86_64-aglsdk-linux/usr/share/cmake/OEToolchainConfig.cmake)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
```
@mkorniichuk position independent code may be setted up with cmake
DOne in #11

mkorniichuk replied:
@LuxoftAKutsan fixed within a89e4042f9a29e6b3ec306046c57bd2bd2e2c4f7

LuxoftAKutsan commented on CMakeLists.txt:
```diff
@@ -224,27 +215,28 @@ add_custom_target(genivi-tarball
-  else()
+  execute_process(COMMAND ${CMAKE_CXX_COMPILER} -fuse-ld=gold -Wl,--version ERROR_QUIET OUTPUT_VARIABLE GOLD_VERSION)
+  if(GOLD_VERSION MATCHES "GNU gold")
     message(INFO " Use gold linker")
```
@mkorniichuk replace INFO to status. There is no such message threshold as INFO

mkorniichuk replied:
@LuxoftAKutsan fixed within a89e4042f9a29e6b3ec306046c57bd2bd2e2c4f7

LuxoftAKutsan commented on tools/infrastructure/docker/AGL/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+    sudo \
+    git \
+    wget \
+    vim \
```
@mkorniichuk remove redundant tools from docker file that should only build SDL

mkorniichuk replied:
@LuxoftAKutsan  fixed within 0e2904f7b83ae3e408736ef42ddbd3d8d833bb68

LuxoftAKutsan reviewed (approved):
:+1:  Please fix minor issues

mkorniichuk committed:
Fix dockerfile for AGL build

* removed toolchain for qnx 6.5.0
* moved agl toolchain to toolchains dir
* removed unnecessary tools from Dockerfile
* adjusted build.sh to in respect to a new path of toolchain file for AGL

mkorniichuk committed:
Fixes according to comments

Minor issues fixed according to comments

mkorniichuk committed:
README.md update: AGL

Updated README.md with section for building for AGL

mkorniichuk committed:
Minor fixes for Docker builds

* removed unnecessary tools for QNX
* minor naming issues fixed

mkorniichuk committed:
Build fixed

fixed build for QNX and AGL for multithreading build, with unit tests and for 
proprietary and external proprietary scenarios

mkorniichuk committed:
Minor fixes

* Apr force rebuild if apr-utils not found
* findAPRUtils fixed according to QNX lib naming

BSolonenko commented on src/3rd_party/apache-log4cxx/AGL.cmake:
```diff
@@ -0,0 +1,43 @@
+  DEPENDS libapr-1
+  DEPENDS apr-util
+  WORKING_DIRECTORY ${LOG4CXX_BUILD_DIRECTORY}
+)
```
@mkorniichuk please add empty line at the end of the file

BSolonenko commented on src/3rd_party/apache-log4cxx/AGL.cmake:
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2018, Ford Motor Company
```
@mkorniichuk Copyright (c) **2019**, Ford Motor Company

vkushnirenko-luxoft commented:
Linux_build

SNiukalov commented on src/3rd_party/boost/Linux.cmake:
```diff
@@ -0,0 +1,32 @@
+    WORKING_DIRECTORY ${BOOST_SOURCE_DIRECTORY} 
+    COMMAND /bin/bash -c \"${INSTALL_COMMAND}\"
+  )"
+)
```
Please add empty line at end file

SNiukalov commented on src/3rd_party/apr/QNX.cmake:
```diff
@@ -51,3 +51,16 @@ set(CONFIGURE_FLAGS
+  COMMAND make include/private/apr_escape_test_char.h
+  COMMAND ${CMAKE_COMMAND} -E copy include/private/apr_escape_test_char.h ${APR_SOURCE_DIRECTORY}/include/private
+  WORKING_DIRECTORY ${APR_BUILD_DIRECTORY}
+)
```
Please add empty line at end file

mkorniichuk committed:
Fix minor issues after rebase

AByzhynar commented on CMakeLists.txt:
```diff
@@ -304,26 +290,22 @@ if(ENABLE_GCOV)
-set(CMAKE_CXX_FLAGS_RELEASE " -DNDEBUG -s -O2")
-set(CMAKE_CXX_FLAGS_DEBUG   " -DDEBUG -g3 -ggdb3")
-
+# TODO(AK): check current OS here
```
@mkorniichuk What do we need `TODO` here for?

mkorniichuk replied:
@AByzhynar removed in ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

AByzhynar commented on src/3rd_party/apache-log4cxx/Linux.cmake:
```diff
@@ -0,0 +1,42 @@
+  DEPENDS libapr-1
+  DEPENDS apr-util
+  WORKING_DIRECTORY ${LOG4CXX_BUILD_DIRECTORY}
+)
```
@mkorniichuk 
Please add an empty line here

mkorniichuk replied:
@AByzhynar fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

AByzhynar commented on src/3rd_party/apache-log4cxx/QNX7.cmake:
```diff
@@ -0,0 +1,66 @@
+  DEPENDS libapr-1
+  DEPENDS apr-util
+  WORKING_DIRECTORY ${LOG4CXX_BUILD_DIRECTORY}
+)
```
@mkorniichuk Check all files for empty line presence at the end

mkorniichuk replied:
@AByzhynar fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

AByzhynar commented on src/3rd_party/apr-util/Linux.cmake:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2014, Ford Motor Company
+# Copyright (c) 2018, Ford Motor Company
```
@mkorniichuk If you decided to change a year - change it to 2019 and in all touched files

mkorniichuk replied:
@AByzhynar fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

AByzhynar commented on src/3rd_party/apache-log4cxx/helpers.sh:
```diff
@@ -0,0 +1,66 @@
+    CURRENT_VERSION=$(cd $LOG4CXX_SOURCE_DIRECTORY && git log --pretty=format:%H -1 .)
+    if [[ $INSTALLED_VERSION == $CURRENT_VERSION ]]; then
+        return 1
+    else
```
@mkorniichuk redundant `else`

mkorniichuk replied:
@AByzhynar fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/apache-log4cxx/CMakeLists.txt:
```diff
@@ -0,0 +1,96 @@
+
+if ((LOG4CXX_FOUND OR NO_REBUILD_3RD_PARTY_LOGGER) AND
+    NOT FORCE_3RD_PARTY_LOGGER)
+
```
@mkorniichuk Extra line?

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/apr-util/AGL.cmake:
```diff
@@ -0,0 +1,43 @@
+  DEPENDS libapr-1
+  DEPENDS expat
+  WORKING_DIRECTORY ${APR_UTIL_BUILD_DIRECTORY}
+)
```
@mkorniichuk Need line in the file end

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/apr-util/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+    USE_SOURCE_PERMISSIONS
+    FILES_MATCHING PATTERN libaprutil-1.so*
+  )
+endif()
```
@mkorniichuk add line

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/apr-util/Linux.cmake:
```diff
@@ -28,46 +28,15 @@
-  USE_SOURCE_PERMISSIONS
-  FILES_MATCHING PATTERN libaprutil-1.so*
-)
+)
```
@mkorniichuk add line

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/apr/CMakeLists.txt:
```diff
@@ -0,0 +1,83 @@
+find_package(APR)
+
+if ((APR_FOUND OR NO_REBUILD_3RD_PARTY_LOGGER) AND
+    NOT FORCE_3RD_PARTY_LOGGER)
```
@mkorniichuk seems like we can have all condition on one line and remove redundant line 39

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/apr/Linux.cmake:
```diff
@@ -61,23 +39,4 @@ add_custom_command(OUTPUT ${APR_BUILD_DIRECTORY}/include/private/apr_escape_test
-  USE_SOURCE_PERMISSIONS
-  FILES_MATCHING PATTERN *.h
-)
+)
```
@mkorniichuk add line

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/apr/QNX7.cmake:
```diff
@@ -51,3 +51,16 @@ set(CONFIGURE_FLAGS
+  COMMAND make include/private/apr_escape_test_char.h
+  COMMAND ${CMAKE_COMMAND} -E copy include/private/apr_escape_test_char.h ${APR_SOURCE_DIRECTORY}/include/private
+  WORKING_DIRECTORY ${APR_BUILD_DIRECTORY}
+)
```
@mkorniichuk add line

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/boost/AGL.cmake:
```diff
@@ -0,0 +1,66 @@
+  "using gcc : agl : x86_64-agl-linux-gcc  -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2 --sysroot=/opt/agl-sdk/6.0.2-corei7-64/sysroots/corei7-64-agl-linux : ")
+
+set(CONFIGURE_COMMAND
+  CC=\"\" ./bootstrap.sh --with-libraries=system,thread,date_time,filesystem --prefix=${3RD_PARTY_INSTALL_PREFIX} COMMAND echo ${BOOST_AGL_PROJECT_CONFIG_JAM} $<SEMICOLON> >> ./project-config.jam)
```
@mkorniichuk Seems like too long line, do we have any constraints of symbols in one line?

mkorniichuk replied:
@ZhdanovP fixed within 160e7a336dbc56913b91e128daabf59f3b69933e

ZhdanovP commented on src/3rd_party/boost/FindBoost.cmake:
```diff
@@ -0,0 +1,283 @@
+    set(Boost_FOUND 0)
+  endif()
+
+  ###########################
```
@mkorniichuk Do we need this fence here?

mkorniichuk replied:
@ZhdanovP removed in ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/boost/QNX7.cmake:
```diff
@@ -28,18 +28,17 @@
-BOOTSTRAP
-./bootstrap.sh --with-toolset=gcc --with-libraries=system,thread,date_time,filesystem --prefix=${3RD_PARTY_INSTALL_PREFIX})
+set(BOOTSTRAP
+  ./bootstrap.sh --with-toolset=gcc --with-libraries=system,thread,date_time,filesystem --prefix=${3RD_PARTY_INSTALL_PREFIX})
```
@mkorniichuk This file contains a lot of long lines, it's seems to be hard to perception.

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/3rd_party/bson/QNX7.cmake:
```diff
@@ -44,4 +46,34 @@
+    WORKING_DIRECTORY ${BSON_LIB_SOURCE_DIRECTORY}
+    COMMAND /bin/bash -c \"${INSTALL_COMMAND}\"
+  )"
+)
```
@mkorniichuk Additional line

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on src/appMain/build_config.txt:
```diff
@@ -1,5 +1,5 @@
 //Policy mode (PROPRIETARY, HTTP or EXTERNAL_PROPRIETARY)
-EXTENDED_POLICY:STRING=PROPRIETARY
+EXTENDED_POLICY:STRING=EXTERNAL_PROPRIETARY
```
@mkorniichuk Will we have EXTERNAL_PROPRIETARY policy by default? Why do you change it?

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented on tools/infrastructure/docker/QNX/entrypoint.sh:
```diff
@@ -0,0 +1,24 @@
+    export HOME=/home/developer
+fi
+
+echo "Welcome to developer_qnx:build!"
```
@alignment alignment if you want

mkorniichuk replied:
@ZhdanovP fixed within ae7d52ab0f457d8f46b1ce7aedaf336c3f9d1933

ZhdanovP commented:
@mkorniichuk Why the "Linux_build" on CI are failed?

mkorniichuk committed:
Fixes according to comments

* moved Find 3rd_party cmake modules to /tools/cmake/modules
* minor fixes according to comments
* Copyright year fix

mkorniichuk committed:
Long line issue

* backslash does not work on cmake version before 3.0

vkushnirenko-luxoft commented:
Linux_build

vkushnirenko-luxoft commented:
Linux_build

vkushnirenko-luxoft commented:
Linux_build

mkorniichuk committed:
Minor fixes

* Fixed broken link in README.md
* Added missing copyright
